nats.go

```diff
@@ -3006,3 +3033,48 @@ func (nc *Conn) TLSRequired() bool {
 	defer nc.mu.Unlock()
 	return nc.info.TLSRequired
 }
+
+// Barrier schedules the given function `f` to all registered asynchronous
+// subscriptions.
+// Only the last subscription to see this barrier will invoke the function.
+// If no subscription is registered at the time of this call, `f()` is invoked
+// right away.
+// ErrConnectionClosed is returned if the connection is closed prior to
+// the call.
+func (nc *Conn) Barrier(f func()) error {
+	nc.mu.Lock()
+	defer nc.mu.Unlock()
+	if nc.isClosed() {
+		return ErrConnectionClosed
+	}
+	nc.subsMu.Lock()
+	defer nc.subsMu.Unlock()
+	// Need to figure out how many non chan subscriptions there are
+	numSubs := 0
```

Comment:
For small numbers should be fine, but should we consider tracking a variable on nc? nc.numAsyncSubs

Reply:
I'd rather leave this for now, avoiding any issue with discrepancies between +1 and -1 when removing a subscription.